sympy/assumptions/handlers/matrices.py

```diff
@@ -155,6 +155,53 @@ def MatrixSlice(expr, assumptions):
         else:
             return ask(Q.invertible(expr.parent), assumptions)
 
+    @staticmethod
+    def MatrixBase(expr, assumptions):
+        if not expr.is_square:
+            return False
+        return expr.rank() == expr.rows
+
+    @staticmethod
+    def MatrixExpr(expr, assumptions):
+        if not expr.is_square:
+            return False
+        return None
+
+    @staticmethod
+    def BlockMatrix(expr, assumptions):
+        from sympy.matrices.expressions.blockmatrix import reblock_2x2
+        if not expr.is_square:
+            return False
+        if expr.blockshape == (1, 1):
+            return ask(Q.invertible(expr.blocks[0, 0]), assumptions)
+        expr = reblock_2x2(expr)
+        if expr.blockshape == (2, 2):
+            [[A, B], [C, D]] = expr.blocks.tolist()
+            if ask(Q.invertible(A), assumptions) == True:
+                invertible = ask(Q.invertible(D - C * A.I * B), assumptions)
+                if invertible is not None:
+                    return invertible
+            if ask(Q.invertible(B), assumptions) == True:
+                invertible = ask(Q.invertible(C - D * B.I * A), assumptions)
+                if invertible is not None:
+                    return invertible
+            if ask(Q.invertible(C), assumptions) == True:
+                invertible = ask(Q.invertible(B - A * C.I * D), assumptions)
+                if invertible is not None:
+                    return invertible
+            if ask(Q.invertible(D), assumptions) == True:
+                invertible = ask(Q.invertible(A - B * D.I * C), assumptions)
+                if invertible is not None:
+                    return invertible
+        return None
```

Comment:
If all 4 blocks are noninvertable can we return False here?

Reply:
Or maybe there's an additional condition on the blocks being square?

Reply:
If all 4 blocks cannot be inverted then it's not correct to imply that the entire matrix is not invertible. It only means that these formulas don't work. Here's an example:
```
a = BlockMatrix([
    [ZeroMatrix(1, 1), Matrix([[2, 3]])],
    [Matrix([[4],[5]]), OneMatrix(2, 2)],
])
```
`block_collapse(a.I)` won't work, but `a.as_explicit().I` works fine, because that matrix is invertible, even if none of the blocks are.